content/blog/next-level-iac-breakpoint-debugging/index.md

```diff
@@ -0,0 +1,317 @@
+    "configurations": [
+        {
+            "name": "Launch Pulumi Program (debug)",
+            "type": "node",
+            "port": 9292,
+            "request": "attach",
+            "preLaunchTask": "pulumi-debug",
+            "continueOnAttach": true,
+            "restart": {
+                "delay": 5000,
+                "maxAttempts": 5
+            },
+            "skipFiles": [
+                "${workspaceFolder}/node_modules/**/*.js",
+                "${workspaceFolder}/lib/**/*.js",
+                "<node_internals>/**/*.js"
+            ]
+        }
+    ]
+}
+```
+{{% notes type="info" %}}
+Let's break this down a bit: 
+
+* `type: node` tells VS Code that we are using the Node runtime
+* `port: 9292` tells VS Code to attach the debugger to port 9292
+* `request: attach` tells VS Code that we want to attach to to the debugger at launch
+* `preLaunchTask: pulumi-debug` tells VS Code to run the task we configured in `tasks.json` before we connect the debugger. This will run our `pulumi-debug.sh` script and then attach to the Node.js runtime 
+* `continueOnAttach: true` tells VS Code to start running our code as soon as we attach to the waiting Node.js runtime (until we reach our first breakpoint that is, where it will pause and return control to the debugger client, aka VS Code)
+* `restart: {...}` configures VS Code to wait 5 seconds and then retry connecting, up to 5 times. This is here in case it takes a while for the Node.js runtime to start running. It's unlikely, but possible!
+* `skipFiles: {...}` sets up some filters of files *not* to try to debug. This includes everything that is part of node's internals, 3rdparty node modules, and other libraries. That way we don't end up deep in someone else's code unexpectedly and can just focus on our Pulumi program.
+* `name` sets a human-readable name which shows up in the drop down box in VS Code when you choose which [launch configuration][vs-code-launch-configs] to run. 
+{{% /notes %}}
+
+Finally, we need to make one more edit to the `pulumi-debug.sh` file we created. Here will will pass the necessary options to the Node.js runtime to set it to run in debug mode. 
+
+Add the following line to the beginning of the file:
+
+```shell {hl_lines=[3]}
+#!/bin/bash
+
+export NODE_OPTIONS='--inspect-brk=127.0.0.1:9292'
+
+pulumi login
+pulumi stack select dev
+pulumi up -f
+```
+{{% notes type="info" %}}
+The `NODE_OPTIONS` environment variable passes options to the Node.js V8 runtime. In this case, we are passing `--inspect-brk` which tells the Node.js runtime to run in debug mode, which will cause it to pause before running the code, until a debugging client connects to `localhost` on port `9292`.
+{{% /notes %}}
+
+Ok, now we should be ready to run our Pulumi program from VS Code using the `F5` key! Let's give it a try. You should see Pulumi running inside of the "Terminal" tab. You will also see output indicating that it is waiting for a debugger to attach: 
```

Comment:
This is point where VS Code is expecting Pulumi to behave differently than it does. It's kind of complicated to get into in the blog post... but essentially, as i explained in a previous comment, the only way this works is to run Pulumi as a background task, which means that VS Code expects it to keep running and produce a series of "problems" while its running which it parses out with a customizable regex-based "problem matcher" configuration to find errors and report them in the Problems tab. but for whatever reason, the problem matcher can't really be configured to work with the type of output Pulumi creates (despite it seeming like it would be possible... i went down a lengthy rabbit hole trying to get this to "just work" and all i was able to do was get partial success and it required an incredibly complex version of `tasks.json` even to get that much working). We could likely create an output mode for the Pulumi CLI that *would* output things in the way that VS Code expects, or create an extension for VS Code that would fix the need to background it, but as it stands, there's not really a way to do it. So, you get this annoying popup box until you click and tell it not to worry about it and to remember it for that task. If I knew where VS Code stored that "remember" i would include steps to configure that also, but I haven't been able to figure where it's stored yet.

Reply:
Actually, I figured out where this is stored. 
In `$HOME/Library/Application\ Support/Code/User/settings.json`
```json
{
...
    "task.problemMatchers.neverPrompt": {
        "shell": true
    },
...
```

But again, not sure if this level of detail is worth including in the article or if folks can just click the checkbox on the prompt and be happy with it. The problem of course, is that this is a global setting for all tasks that are `shell` tasks, so this would apply to everything, not just our example here. This might be too much to tell users to configure to follow along with our example as it will hide issues with problem matchers in future `shell` tasks.